torch/optim/lr_scheduler.py

```diff
@@ -0,0 +1,206 @@
+import numpy as np
+from bisect import bisect_right
+from torch.optim.optimizer import Optimizer
+
+
+class LambdaLR(object):
+    def __init__(self, optimizer, base_lr, lr_lambda):
+        self.optimizer = optimizer
+        self.base_lr = base_lr
+        self.lr_lambda = lr_lambda
+
+    def step(self, epoch):
+        for param_group in self.optimizer.param_groups:
+            param_group['lr'] = self.base_lr * self.lr_lambda(epoch)
+
+
+class GroupLambdaLR(object):
+    def __init__(self, optimizer, base_lrs, lr_lambdas):
+        self.optimizer = optimizer
+        self.base_lrs = base_lrs
+        self.lr_lambdas = lr_lambdas
+
+    def step(self, epoch):
+        for param_group, base_lr, lr_lambda in zip(
+                self.optimizer.param_groups,
+                self.base_lrs, self.lr_lambdas):
+            param_group['lr'] = base_lr * lr_lambda(epoch)
+
+
+class StepLR(LambdaLR):
+    """Set the learning rate to the base_lr decayed by gamma
+    every step_size epochs.
+
+
+    Example:
+        >>> # lr = 0.05     if epoch < 30
+        >>> # lr = 0.005    if 30 <= epoch < 60
+        >>> # lr = 0.0005   if 60 <= epoch < 90
+        >>> # ...
+        >>> scheduler = StepLR(optimizer, base_lr=0.05, gamma=0.1, step_size=30)
+        >>> for epoch in range(100):
+        >>>     scheduler.step(epoch)
+        >>>     train(...)
+        >>>     validate(...)
+    """
+
+    def __init__(self, optimizer, base_lr=0.1, gamma=0.1, step_size=30):
+        super(StepLR, self).__init__(optimizer, base_lr,
+                                     lambda epoch: gamma ** (epoch // step_size))
+
+
+class MultiStepLR(LambdaLR):
+    """Set the learning rate to the base_lr decayed by gamma
+    once the number of epoch reaches one of the milestones.
+
+
+    Example:
+        >>> # lr = 0.05     if epoch < 30
+        >>> # lr = 0.005    if 30 <= epoch < 80
+        >>> # lr = 0.0005   if epoch >=80
+        >>> scheduler = MultiStepLR(optimizer, base_lr=0.05, gamma=0.1, milestones=[30,80])
+        >>> for epoch in range(100):
+        >>>     scheduler.step(epoch)
+        >>>     train(...)
+        >>>     validate(...)
+    """
+
+    def __init__(self, optimizer, base_lr=0.1, gamma=0.1, milestones=(10, 20, 30)):
```

Comment:
IMO, separating them is more natural and easy to understand. Merging these might cause confusion. 
Sort of agree with the verifying part. Is the point about letting the user know their input is wrong without changing it?